LibOS/shim/src/fs/proc/thread.c

```diff
@@ -491,54 +490,125 @@ static int proc_thread_maps_open (struct shim_handle * hdl,
     if (!thread)
         return -ENOENT;
 
-    int size = DEFAULT_BUFFER_SIZE;
-    char * strbuf = malloc(size);
-    int ret = 0, len = 0;
+    size_t count = DEFAULT_VMA_COUNT;
+    struct shim_vma_val * vmas = malloc(sizeof(struct shim_vma_val) * count);
 
-    if (!strbuf) {
+    if (!vmas) {
         ret = -ENOMEM;
         goto out;
     }
 
-retry:
-    ret = dump_all_vmas(thread, strbuf, size);
+retry_dump_vmas:
+    ret = dump_all_vmas(vmas, count);
 
     if (ret == -EOVERFLOW) {
-        char * newbuf = malloc(size * 2);
-        if (!newbuf) {
+        struct shim_vma_val * new_vmas
+                = malloc(sizeof(struct shim_vma_val) * count * 2);
+        if (!new_vmas) {
             ret = -ENOMEM;
             goto err;
         }
-        free(strbuf);
-        strbuf = newbuf;
-        size *= 2;
-        goto retry;
+        free(vmas);
+        vmas = new_vmas;
+        count *= 2;
+        goto retry_dump_vmas;
```

Comment:
I think moving this part into a function (in a header) might make this code less readable. I am fine with having some similar coding patterns in different places.

Reply:
This IMO is not a "coding pattern", but a small, copy-pasted algorithm, which is complicated enough to be extracted to a separate function. Some reasons:
* The reader now has to read unnecessary implementation details of something totally unrelated to the function he reads. This could be implemented as just one line:
```C
ret = alloc_and_dump_all_vmas(&vmas);
```
* This is likely that we'll want to do some changes (think e.g. about adjusting that `*2` factor), it's likely that the copied parts will diverge from each other.
* The code is almost identical in all places, so it's not time consuming / problematic to do it.

Reply:
I don't think moving a piece of one-time code into a separate function will have any benefits. Here are some disputes to the reasons you brought up:
  * Hiding some logic inside some function names that we "assume" to be straightforward is even worse than letting the readers see the raw code. Take your example as an example: what does it mean for "alloc_and_dump_all_vmas"? Do you mean "dump" as "copying" or just "trashing"? Is the original vmas freed? How big is the vmas buffer enlarged? Readers still have to jump to the function to read what's inside it. As a result, it is not making anything better.

  * Yes, we may want to adjust the *2 factor, but on a case-by-case basis. 

  * As you mentioned, the code is "almost" identical. "Almost".  Sometimes, there is a need of copying the buffer content to the new buffer. Sometimes, we need the buffer size for later user. All these problems cause complication to the function.

Reply:
There are two key differences: (1) How reusable exactly the function is. (2) Whether the function secretly allocating new buffer without user's knowledge (which usually leads to memory leak).

For the *2 factor, I think I misunderstood what you said. Yes, it's unlikely that we will change this factor, except sometimes we have more insight about how fast we should grow the buffer.